Move responsability of loading profile to `ProfileParamType`

sphuber opened the pull request:
Currently, the loading of the profile is done in the body of the `verdi`
command group itself. By moving this to the `ProfileParamType`, it
becomes easy for external scripts to also provide a `--profile` flag to
run the script for a profile other than the default.

ltalirz commented on aiida/cmdline/commands/cmd_verdi.py:
```diff
@@ -14,36 +14,27 @@
     if ctx.obj is None:
         ctx.obj = AttributeDict()
 
     config = get_config(create=True)
```
just wondering - the profile option will already have loaded the config.
Should we be passing the config in the context instead?

sphuber replied:
This would require us to create the `ctx.obj` in the option type as well, binding it tightly to this concept. On top of that, we still have to keep it here, because if the `--profile` is not explicitly used, since there is no default, the `ProfileParamType.convert` won't even be called. So I prefer to just load the config here. If it has already been called before, this will just return the same instance, since it is a global and the call will be very cheap

ltalirz commented on aiida/cmdline/commands/cmd_verdi.py:
```diff
@@ -14,36 +14,27 @@
         ctx.obj = AttributeDict()
 
     config = get_config(create=True)
+    profile = get_profile()
```
why not use the profile that is returned by the option?

sphuber replied:
Because the option will return `None` if no specific profile was specified. However, other commands depend on `ctx.obj.profile` to contain the profile that is to be used.

ltalirz replied:
I see - is this the behavior we want, though?
If somebody else wants to reuse the `--profile` option, I guess it would be natural to want to get the default profile, if non is specified.
I.e. can't we make the default behavior to load the default profile?

sphuber replied:
I thought about this and think not (but not 100% sure), because it won't be possible to know whether the `-p` flag was explicitly set or the default was used, which is what we set in `ctx.obj.profile_option_used`. This boolean is now used for example in the `get_default_profile` function which is used by `verdi profile show` as callback for the profile argument. Let's say default profile is `django`. If I call `verdi profile show` it will show that of `django`. But if I call `verdi -p sqla profile show` it will still show details of django, if it weren't for the `ctx.obj.profile_option_used`.

This is currently used only in this callback function. I will try now if I can get rid of it and have the `PROFILE` option to use a default

sphuber commented:
@ltalirz I have pushed another commit with quite a few additional changes. After some more thought I agree that giving the `PROFILE` option a default, which resolves to the default profile makes sense. As a result this makes the `cmd_verdi` implementation a lot cleaner. The down side is that the `profile_option_used` context attribute can no longer be defined. However, it was only being used in the `get_default_profile` function, which was only used by `verdi profile show`. I changed this callback into a plain function, which is now used as the default lambda for the `PROFILE` option. The `verdi profile show` now get's the correct profile simply from the `ctx.obj.profile` so we no longer need the `profile_option_used`.

On top of this, I found out that the implementation of `load_profile` was always triggering the loading of the backend. With us moving the loading of the profile to the `PROFILE` option, this was now always causing the backend to be loaded, even when just calling `verdi`. I have fixed this issue by moving the related code to the `Manager._load_backend` where it will be called lazily as it should.

The scope of the PR has changed quite a bit, but I wanted to go through the review first. Once we are happy I will rebase and squash and rephrase the commit message.

sphuber commented on aiida/cmdline/utils/defaults.py:
```diff
@@ -17,33 +17,24 @@
+def get_default_profile():  # pylint: disable=unused-argument
     """Try to get the default profile.
 
     This should be used if the default profile should be returned lazily, at a point for example when the config
```
I guess it is missing that it can return `None` if no default profile is set. But if you are referring to the explanation of where this should be used, that is still valid.

ltalirz replied:
I meant: this says 

> this should be used if the default profile should be returned lazily [...]Otherwise, the preference should go to calling `get_config` to load the actual config and using `config.default_profile_name`

However, now it returns the profile *name*.
I'm not sure there's really any preference for manually calling`get_config` anymore.

sphuber replied:
I will update it to be more specific, but it does reference `config.default_profile_name`. The usage rule still applies. If one needs to know the default profile anywhere in application code, one should always go through the `config` instance. This function should only be used when this is not possible, for example when defining a default for the command line option

sphuber commented on aiida/cmdline/utils/defaults.py:
```diff
@@ -17,33 +17,24 @@
 
-def get_default_profile(ctx, param, value):  # pylint: disable=unused-argument
+def get_default_profile():  # pylint: disable=unused-argument
     """Try to get the default profile.
```
Done

sphuber commented on aiida/common/log.py:
```diff
@@ -161,7 +162,7 @@ def configure_logging(with_orm=False, daemon=False, daemon_log_file=None):
     from logging.config import dictConfig
 
-    config = evaluate_logging_configuration(LOGGING)
+    config = evaluate_logging_configuration(copy.deepcopy(LOGGING))
```
The name is correct in a sense that the `LOGGING` dictionary contains lambdas that need to be evaluated lazily. This is done by `evaluate_logging_configuration`. In that function I recursively iterate the mapping, creating copies of nested dictionaries exactly because the original should not be touched. However, it does not deal with sequences, so the nested lists in the evaluated copy will be references to the original. This leads to problems. I could have tried to add a clause for `isinstance(value, collections.Sequence)` but passing a deepcopy of the original structure is bullet-proof

ltalirz replied:
Ah ok - I thought "evaluate" was meant more in the sense of "analyze" than "execute".

sphuber commented on aiida/manage/configuration/__init__.py:
```diff
@@ -53,19 +52,15 @@ def load_profile(profile=None):
-    manager = get_manager()
-    manager.unload_backend()
-    reset_manager()
+    configure_logging()
```
`configure_logging` actually does not use information from the profile. The information is "inserted" lazily by using the lambdas in the `LOGGING` dictionary. These are evaluated when `configure_logging` is called, but if you look at the implementation it does not use any information from the profile. It does not even get it. The `configure_repository` on the other hand does require information from the profile

sphuber commented on aiida/manage/manager.py:
```diff
@@ -111,6 +112,10 @@ def _load_backend(self, schema_check=True):
 
+        # Reconfigure the logging with `with_orm=True` to make sure that profile specific logging configuration options
+        # are taken into account and the `DbLogHandler` is configured.
+        configure_logging(with_orm=True)
```
Yes it will. The idea is that in principle we always want the logging to be configured as soon as a profile is loaded but we do not want to configure the `DbLogHandler` because that requires loading the database. This needs to be done then when the backend is loaded. I don't see any way around this. Unless we remove the requirement that the logging is configured as soon as a profile is loaded, which seems undesirable, but might be ok, I am not sure

ltalirz replied:
I see. Perhaps just adding a bit of documentation might help for the moment, i.e.
 * extending the comment where `configure_logging` is used without `with_orm`
 * documenting the `with_orm` parameter inside `configure_logging` (seems to be missing).

sphuber replied:
will do

sphuber commented on aiida/cmdline/commands/cmd_profile.py:
```diff
@@ -52,9 +52,13 @@ def profile_list():
+@click.pass_context
+def profile_show(ctx, profile):
     """Show details for PROFILE or, when not specified, the default profile."""
+    if profile is None:
```
well in `setdefault` I make it required, because without value the command does not make sense. So there I am guaranteed that `profile is not None`. However, `verdi profile show` does not require a profile, so the `profile` will be `None` if not specified. However, in that case `ctx.obj.profile` _should_ have been defined in `verdi` itself. But just to be sure I keep the check in case no profile exists yet, for example.

ltalirz replied:
> However, in that case ctx.obj.profile should have been defined in verdi itself. But just to be sure I keep the check in case no profile exists yet, for example.

Not quite sure I follow the logic here - if no profile exists, wouldn't `ctx.obj.profile` be `None` as well?

sphuber replied:
Yes, exactly, since `ctx.obj.profile` is not guaranteed to contain a profile, for example in the case when no profiles exist yet, I don't want `verdi process show` to except, so therefore I have to check for `None`

sphuber commented:
@ltalirz let me know if you are happy with my answers or you think something has to be changed. I haven't updated yet because I wanted to wait for your response first

ltalirz commented:
@sphuber Thanks - I'm on holiday this week, so replies might be a bit more sporadic than usual ;-)

sphuber commented:
> @sphuber Thanks - I'm on holiday this week, so replies might be a bit more sporadic than usual ;-)

Oh shit, sorry, I didn't realize. I will quickly update this PR with the last few changes to docstrings and then I will leave you in peace!

sphuber commented:
@ltalirz this should be good to go
